docs/programming-guide.md

```diff
@@ -1022,6 +1022,27 @@ for details.
 </tr>
 </table>
 
+### Shuffle operations
+
+Certain operations within Spark trigger an operation known as the shuffle. The shuffle is Spark's mechanism for re-distributing data so data with the same key becomes co-located after a shuffle. 
```

Comment:
I think a shuffle is not just about collecting data by key. For example a repartitioning can cause a shuffle. Personally I'd say that some operations need to redistribute data so that it is grouped differently into partitions, which typically means rearranging and copying data across executors or machines.